### DISCUSSION TITLE
VanCone Features

### DISCUSSION THREAD

**Discussion by yahia-berashish** [vanjs-org/van#185] [Q&A]
Hello again. 
Very happy to finally see a full SPA implementation @b-rad-c 
But I have some questions, mainly if VanCone supports the following features, and if not, how can they be implemented: 
- TypeScript, I haven't noticed a TS file in the code
- Component routing: declaring routes with components, like React Router DOM
- Slot/Outlet to create layouts, to unify the UI between sub-routes of the same parent route

I would be happy to help with anything 👋🏼

**Comment by b-rad-c** [comment]
Hey @yahia-berashish - glad you like it!

TypeScript - I'm definitely interested in adding support for this. Of your feature requests this is the one I'm most interested in, I have other projects taking up bandwidth if you're interested in helping this would be a good one! It would be nice to get it done now before adding more features.

For Component routing, are you talking about [something like this](https://reactrouter.com/en/main/components/routes)? It seems like this would also go with the slot/outlet feature for sub-routes. I hadn't initially thought about either of these two ideas but I'm kind of intrigued now. I'm gonna sleep on it and see if anything comes to me. Seems like it could be possible to come up with a lightweight API that could handle this. If you have any ideas on what an API would look like feel free to put out some ideas!

**Reply by yahia-berashish** [reply]
Adding TypeScript support is very simple, just creating an `spa.d.ts` file in the `src/` directory and populating it with some types will do the job, and I will be happy to help with that, hopefully, I will be able to open a PR to add the file. 
As for the routing components, it should be something like the [`vanjs-router`](https://github.com/iuroc/vanjs-router) implementation, and I think it can be easily implemented by creating some functions that return route objects, it is just syntactic sugar basically. 
The slot functionality is more complicated, it is not related to the router components though, I tried building a simple implementation of it but ended up needing some kind of a global storage, this can be achieved by multiple methods like the `window` object or a global variable for example, but it got me eventually to think about building a context API for handling global state with @Hunter-Gu on this issue: #161, if you think you can help with finishing that (we made multiple implementations but they all have flaws) it can be integrated into VanCone and it can greatly enhance the router implementation as well as state handling in VanJS generally, and we will appreciate the help of course!

**Comment by b-rad-c** [comment]
I just pushed an update yesterday that adds another property to the route definition:

```javascript
{
      path: "/user/:userId",
      backend: "/secure/user/:userId",      // new property
      name: "user",
      title: "VanJS Example | User",
      callable: async () => import('./app/pages/user')
}
```

A potential route API could look something like this:

```javascript
router(
    route("user", "/user/:userId", userPage, {title: "VanJS Example | User", back: "/secure/user/:userId"}),
    route("item", "/items/:itemID", itemPage, {title: "VanJS Example | Item"}),
    route("home", "/", homePage, {title: "VanJS Example"}),
    route("notFound", ".*", homePage, {title: "VanJS Example | Not Found"})
)
```

`title` and `backend` are in an object because they are both optional params.

For nested child routes you could have something like this:

```javascript

router(
    route("user", "/user/:userId", userPage, {title: "VanJS Example | User", back: "/secure/user/:userId"},
        route("user-inbox", "/inbox", userInbox),
        route("user-feed", "/feed", userFeed),
        route("user-gallery", "/gallery", userGallery)
    ),
    route("item", "/items/:itemID", itemPage, {title: "VanJS Example | Item"}),
    route("home", "/", homePage, {title: "VanJS Example"}),
    route("notFound", ".*", homePage, {title: "VanJS Example | Not Found"})
)

```

If you look in the [spa.js](https://github.com/b-rad-c/van/blob/6271f4927dc57fe64c4435bc7a2ec7e16c4e25b1/addons/van_cone/src/spa.js#L125) file in the `createCone` function it loops through all of the routes and dynamically creates the handler for each. I think it would be possible to pass in the the child element (for example `userInbox` above) and have the route be able to consume it the way it can consume route params or url parameters without adding an additional global state.

```javascript
const userPage = (params, query, context, outlet) => {

  /* our parent component, a search page may have passed in preloaded information via navLink, if not, we'll fetch it */
  const userInfo = (context) ? context : fetchUserInfo(params.id)

  return div(
    p("Hi I am: " + userInfo.name),
    p("some query string data: " + query.data),
    outlet()
  );
}

```

I like the idea of generating routes this way but there are a couple other change I want to make to the repository before adding in this because those changes could affect this API. However if you're motivated to work on it I would happily accept a pull request! It's probably easiest to state without doing the nested routes and then build up from there.

Thanks!

**Reply by yahia-berashish** [reply]
That looks promising, I will try to take a look at it as well. 
Will try and finish the TypeScript part first though because it is the simplest task right now, I will try to update the code and send a PR to your repo as soon as possible.
